apps/functions-subscription/src/adapters/azure/cosmosdb/activation.ts

```diff
@@ -0,0 +1,119 @@
+import * as TE from 'fp-ts/TaskEither';
+import * as E from 'fp-ts/Either';
+import { pipe } from 'fp-ts/lib/function';
+import {
+  BulkOperationType,
+  Database,
+  OperationInput,
+  OperationResponse,
+} from '@azure/cosmos';
+import {
+  ActivationRequestItem,
+  ActivationRequestItemCodec,
+  ActivationResult,
+  ActivationConsumer,
+  BaseActivationItemCodec,
+} from '../../../domain/activation';
+import { decodeFromList } from './decode';
+
+const makeActivationJobPatchOperation = <T extends BaseActivationItemCodec>(
+  obj: T,
+  requests: readonly ActivationRequestItem[],
+  propertyToUpdate: keyof T,
+): OperationInput => ({
+  id: obj.id,
+  operationType: BulkOperationType.Patch,
+  resourceBody: {
+    operations: [
+      {
+        op: 'incr',
+        path: `/${propertyToUpdate.toString()}`,
+        value: requests.length,
+      },
+    ],
+  },
+});
+
+const makeActivationRequestPatchOperation =
+  <T extends BaseActivationItemCodec>(propertyToUpdate: keyof T) =>
+  (obj: T): OperationInput => ({
+    id: obj.id,
+    ifMatch: obj._etag,
+    operationType: BulkOperationType.Patch,
+    resourceBody: {
+      operations: [
+        {
+          op: 'replace',
+          path: `/${propertyToUpdate.toString()}`,
+          value: true,
+        },
+      ],
+    },
+  });
+
+export const makeActivationCosmosContainer = (
+  db: Database,
+): ActivationConsumer => {
+  const container = db.container('activations');
+  return {
+    fetchActivationRequestItemsToActivate: (job) =>
+      pipe(
+        TE.tryCatch(
+          () =>
+            container.items
+              .query({
+                query:
+                  'SELECT * FROM c WHERE c.trialId = @trialId AND c.type = "request" AND c.activated = false OFFSET 0 LIMIT @limit',
+                parameters: [
+                  {
+                    name: '@trialId',
+                    value: job.trialId,
+                  },
+                  {
+                    name: '@limit',
+                    value: job.usersToActivate - job.usersActivated,
```

Comment:
**issue**: This query could be quite costly in terms of Request Units (RU), as it potentially fetches many activation requests. What do you think to restrict to only the request we are going to handle?